lib/auth/grpcserver.go

```diff
@@ -1330,14 +1330,18 @@ func (g *GRPCServer) UpsertApplicationServer(ctx context.Context, req *proto.Ups
 	server := req.GetServer()
 	app := server.GetApp()
 
-	// Only allow app servers with Okta origins if coming from an Okta role. App servers sourced from
+	// Only allow app servers with Okta subkinds/origins if coming from an Okta role. App servers sourced from
 	// Okta are redirected differently which could create unpredictable or insecure behavior if applied
 	// to non-Okta apps.
-	hasOktaOrigin := server.Origin() == types.OriginOkta || app.Origin() == types.OriginOkta
-	if builtinRole, ok := auth.context.Identity.(authz.BuiltinRole); !ok || builtinRole.Role != types.RoleOkta {
-		if hasOktaOrigin {
-			return nil, trace.BadParameter("only the Okta role can create app servers and apps with an Okta origin")
+	if server.GetSubKind() == types.SubKindOktaApp || server.Origin() == types.OriginOkta ||
+		app.GetSubKind() == types.SubKindOktaApp || app.Origin() == types.OriginOkta {
+		if builtinRole, ok := auth.context.Identity.(authz.BuiltinRole); !ok || builtinRole.Role != types.RoleOkta {
+			return nil, trace.BadParameter("only the Okta role can create Okta app servers")
 		}
+		server.SetSubKind(types.SubKindOktaApp)
+		server.SetOrigin(types.OriginOkta)
+		app.SetSubKind(types.SubKindOktaApp)
+		app.SetOrigin(types.OriginOkta)
 	}
```

Comment:
I guess my goal here was to prevent non-Okta roles from adding in Okta origin labels.

I would rather maintain the origin label for UI purposes. At the moment there's no other way to discern what apps belong to Okta and what apps don't without some UI work.